File: 44/44_xjf.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        if p == '': return s == ''
        vis = [[False for j in range(len(p) + 1)] for i in range(len(s) + 1)]
        def dfs(si, pi):
            if vis[si][pi]: return False
            vis[si][pi] = True
            if pi == len(p): return si == len(s)
            if si == len(s): return (dfs(si, pi + 1) if p[pi] == '*' else False)
            if p[pi].isalpha() and s[si] != p[pi]: return False
            if p[pi] != '*': return dfs(si + 1, pi + 1)
            else:
                for ssi in range(si, len(s) + 1):
                    if dfs(ssi, pi + 1): return True
            return False
        return dfs(0, 0)

File: 44/44_xjf.java
//这题和第十题正则表达式一样, 比第十题还简单一点
class Solution {
    boolean vis[][];
    public boolean isMatch(String s, String p) {
        vis = new boolean[s.length() + 3][p.length() + 3];
        return isMatch(s, p, 0, 0);
    }

    private boolean isMatch(String s, String p, int si, int pi) {
        if (vis[si][pi]) return false;
        for (;pi < p.length(); ++pi) {
            if (p.charAt(pi) == '?' || (si < s.length() && p.charAt(pi) == s.charAt(si))) ++si;
            else if (p.charAt(pi) == '*') {
                while (si < s.length()) {
                    if (isMatch(s, p, si, pi + 1)) return true;
                    vis[si++][pi + 1] = true;
                }
            }
            else return false;
        }
        return si == s.length();
    }
}